### DISCUSSION TITLE
No available Views for Resource - extensions installed and activated.

### DISCUSSION THREAD

**Discussion by DiegoDVillacreses** [ckan/ckan#7412] [Q&A]
## Describe the bug
No available Views for any Resource even after installation of `Plotly` and `ckanext-visualize`. My plugins are:
`ckan.plugins = image_view datatables_view datastore visualize plotly`

### Steps to reproduce
I carefully followed your guide https://docs.ckan.org/en/2.10/maintaining/installing/install-from-source.html. Installed all pre-requesites, datastore, and FileStore. All under a local Ubuntu and all configurations point to `localhost`. In any case here is a link to a copy of my `/etc/ckan/default/ckan.ini`: https://drive.google.com/file/d/177FdZH8SLcJ-FpA3lmTNy3Thmtc7slvy/view?usp=share_link

CKAN server seems to work right but is not capable to add any View even before installation of `Plotly` and `ckanext-visualize`. 

For example when only using: `ckan.plugins = datatables_view`, I was also unable to add any view. 

### Additional details
I am using Ubuntu 20.04 as suggested:

```
 user@ubuntu:/usr/lib/ckan/default/src/ckan$ cat /etc/os-release
NAME="Ubuntu"
VERSION="20.04.5 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.5 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal
```

Here are the Python packages installed within my CKAN env:

```
(default) user@ubuntu:/usr/lib/ckan/default/src/ckan$ pip list
Package               Version     Editable project location
--------------------- ----------- ------------------------------
alembic               1.8.1
Babel                 2.10.3
backports.zoneinfo    0.2.1
Beaker                1.11.0
bleach                5.0.1
blinker               1.5
certifi               2021.10.8
charset-normalizer    2.0.12
ckan                  2.10.0      /usr/lib/ckan/default/src/ckan
ckanext-plotly        0.0.1       /home/user/ckanext-plotly
ckanext-visualize     1.0.0
click                 8.1.3
Deprecated            1.2.13
dominate              2.7.0
feedgen               0.9.0
Flask                 2.0.3
Flask-Babel           1.0.0
Flask-Login           0.6.1
flask-multistatic     1.0
Flask-WTF             1.0.1
greenlet              1.1.2
idna                  3.3
importlib-metadata    4.11.3
importlib-resources   5.10.2
itsdangerous          2.1.1
Jinja2                3.1.2
lxml                  4.9.1
Mako                  1.2.2
Markdown              3.4.1
MarkupSafe            2.1.1
mypy                  0.971
mypy-extensions       0.4.3
natsort               7.1.1
nose                  1.3.7
packaging             21.3
passlib               1.7.4
pip                   23.0
pkg_resources         0.0.0
polib                 1.1.1
psycopg2              2.9.3
PyJWT                 2.4.0
pyparsing             3.0.7
pysolr                3.9.0
python-dateutil       2.8.2
python-magic          0.4.27
pytz                  2021.3
pytz-deprecation-shim 0.1.0.post0
PyUtilib              6.0.0
PyYAML                6.0
redis                 4.1.4
requests              2.28.1
rq                    1.11.0
setuptools            44.1.0
simplejson            3.17.6
six                   1.16.0
SQLAlchemy            1.4.41
sqlalchemy2-stubs     0.0.2a27
sqlparse              0.4.2
tomli                 2.0.1
typing_extensions     4.3.0
tzdata                2022.1
tzlocal               4.2
urllib3               1.26.9
watchdog              2.1.6
webassets             2.0
webencodings          0.5.1
Werkzeug              2.0.3
wrapt                 1.14.0
WTForms               3.0.1
zipp                  3.7.0
zope.interface        5.4.0
```
After initializing CKAN I get the following information from terminal:

```
(default) user@ubuntu:/usr/lib/ckan/default/src/ckan$ ckan -c /etc/ckan/default/ckan.ini run
2023-02-16 18:23:55,010 INFO  [ckan.cli] Using configuration file /etc/ckan/default/ckan.ini
2023-02-16 18:23:55,010 INFO  [ckan.config.environment] Loading static files from public
2023-02-16 18:23:55,531 INFO  [ckan.config.environment] Loading templates from /usr/lib/ckan/default/src/ckan/ckan/templates
2023-02-16 18:23:56,212 INFO  [ckan.config.environment] Loading templates from /usr/lib/ckan/default/src/ckan/ckan/templates
2023-02-16 18:23:56,630 WARNI [ckan.config.middleware.flask_app] Extensions are excluded from CSRF protection! We allow extensions to run without CSRF protection but it will be forced future releases. Read the documentation for more information on how to add CSRF protection to your extension.
2023-02-16 18:23:56,652 INFO  [ckan.cli.server] Running CKAN on http://localhost:5000
2023-02-16 18:23:57,525 INFO  [ckan.cli] Using configuration file /etc/ckan/default/ckan.ini
2023-02-16 18:23:57,526 INFO  [ckan.config.environment] Loading static files from public
2023-02-16 18:23:57,794 INFO  [ckan.config.environment] Loading templates from /usr/lib/ckan/default/src/ckan/ckan/templates
2023-02-16 18:23:58,230 INFO  [ckan.config.environment] Loading templates from /usr/lib/ckan/default/src/ckan/ckan/templates
2023-02-16 18:23:58,477 WARNI [ckan.config.middleware.flask_app] Extensions are excluded from CSRF protection! We allow extensions to run without CSRF protection but it will be forced future releases. Read the documentation for more information on how to add CSRF protection to your extension.
2023-02-16 18:23:58,499 INFO  [ckan.cli.server] Running CKAN on http://localhost:5000
```

**Comment by lgmatabh** [comment]
My case is similar.

The CKAN 2.10 server seems to work fine but is not able to display recline_view;

My installation is the basics of the official documentation with only the xloader installation (datastore and recline_view plugins enabled). **Language is pt_BR.**

ckan.plugins = **xloader** datastore recline_grid_view text_view image_view **recline_view** pdf_view activity
It does not find the view and displays the error **"There are no views created for this resource yet."**

The "loading-dialog" class goes into a loop.

The resource and view were created.

1) SELECT id FROM public.resource
ORDER BY id ASC

"id"
"ad34cc39-1782-4248-b7a1-3bace43a3fd7"

2) SELECT * FROM public.resource_view
ORDER BY id ASC

"id" "resource_id" "title" "description" "view_type" "order" "config"
"3ce4f0df-1c0e-41f9-899a-2b0780727f9e" "ad34cc39-1782-4248-b7a1-3bace43a3fd7" "test" "recline_view" 0

3) select count(*) from public."ad34cc39-1782-4248-b7a1-3bace43a3fd7";

13181

4) SELECT * FROM public._table_metadata;

"_id" "name" "oid" "alias_of"
"7f61096d2b943615" "ad34cc39-1782-4248-b7a1-3bace43a3fd7" "22637"

**Comment by amercader** [comment]
@DiegoDVillacreses I'm not familiar with the plotly or ckanext-visualize extensions but at least `datatables_view` requires the resource data to be in the DataStore to be rendered. From your plugins it doesn't look like your are using a tool like datapusher, xloader or datapusher+ to get the data automatically into the Datastore after uploading it so that might be the issue.
To give more context, each resource view plugin has a `can_view()` method that determines if the view will be avbailable for a particular resource. This checks things like the format of the resource or whether it is in the DataStore. You can try to check the source of the plugins you mention to see what you are missing

@lgmatabh in your case I'd check if the [`ckan.views.default_views`](https://docs.ckan.org/en/latest/maintaining/configuration.html#ckan-views-default-views) setting includes the necessary plugins. I'd also recommend replacing the recline_view with datatables_view as the former is no longer supported

**Reply by lgmatabh** [reply]
Hi @amercader,

I configured only these plugins below and uploaded again.

ckan.plugins = **xloader datastore datatables_view activity**

ckan.views.default_views = **datatables_view text_view pdf_view**

It keeps showing this message: **"There are no views created for this resource yet."**

**The iframe is blank**, follow the source code of the attached iframe.

Upload and views created successfully for datatables_view.

1) SELECT id FROM public.resource
ORDER BY id ASC

"ad34cc39-1782-4248-b7a1-3bace43a3fd7"

2) SELECT * FROM public.resource_view
ORDER BY id ASC

"id"	"resource_id"	"title"	"description"	"view_type"	"order"	"config"
"430f73f0-a102-4df7-947c-0a571f378bf0"	"ad34cc39-1782-4248-b7a1-3bace43a3fd7"	"Tabela"		"datatables_view"	1	"{""date_format"": ""llll"", ""ellipsis_length"": 100, ""filterable"": true, ""responsive"": false, ""show_fields"": [""_id"", ""Id"", ""DataAssinatura"", ""TipoArquivo"", ""Assinante"", ""Aplicativo"", ""Sucesso""]}"

3) select count(*) from public."ad34cc39-1782-4248-b7a1-3bace43a3fd7";
13181

4) SELECT * FROM public._table_metadata

"_id"	"name"	"oid"	"alias_of"
"7f61096d2b943615"	"ad34cc39-1782-4248-b7a1-3bace43a3fd7"	"31074"	


[iframe_preview_datatables.txt](https://github.com/ckan/ckan/files/10796537/iframe_preview_datatables.txt)

Thanks.

**Comment by lgmatabh** [comment]
Hi @amercader,

I did the same basic installation of CKAN 2.9.8 and 2.10 from the official documentation in different directories, with the same ckan.ini file and using the same ckan_default and datastore_default databases

In VERSION 2.9.8, the datatables_view is ok and in 2.10 it does not display the data, leaving the Iframe of the page blank in version 2.10

**There is a bug or need for some additional configuration in version 2.10.**

both with:

ckan.plugins = xloader datatables_view datastore text_view image_view recline_view
ckan.views.default_views = datatables_view image_view text_view recline_view

**Reply by osoaresjunior** [reply]
it's not bug, it´s redirect

**Comment by lgmatabh** [comment]
Hi  @amercader,,

I debugged in chrome and found this error regarding blank page displayed. Is there any sense?

**Error:**

3ad3c60e_datatablesview.js:26916 Uncaught TypeError: Cannot read properties of undefined (reading 'toLowerCase')
    at 3ad3c60e_datatablesview.js:26916:44
(anônimo) @ 3ad3c60e_datatablesview.js:26916

...

26901 /* global $ jQuery gdataDict gresviewId */
26902 
26903 // global vars used for state saving/deeplinking
26904 let gsavedPage
26905 let gsavedPagelen
26906 let gsavedSelected
26907 // global var for current view mode (table/list)
26908 let gcurrentView = 'table'
26909 // global var for sort info, global so we can show it in copy/print
26910 let gsortInfo = ''
26911 // global vars for filter info labels
26912 let gtableSearchText = ''
26913 let gcolFilterText = ''
26914 
26915 let datatable
**26916 const gisFirefox = navigator.userAgentData.toLowerCase().indexOf('firefox') > -1**
26917 let gsearchMode = ''
26918 let gstartTime = 0
26919 let gelapsedTime
26920 
26921 // HELPER FUNCTIONS

...

Thanks.

**Reply by osoaresjunior** [reply]
So I solved this in ckan 2.10.4 in a simple way, what happens is that if you are using https and a direct redirection is being made from http to https, the data will not be viewed, try to see if you can see it without https, that is, if the data is viewed via http, use Inmsonia or another to view it using the get method. If it is being viewed, simply change the view in the ckan.ini file by placing https in ckan.site_url = https://xxxxx and do http to https redirection in nginx or apache
